Apparent regression after PR #263: race in hole fill

This appears to be a regression after the merge of #263. In the parallel version of QuickCheck test for the SMRMap, there's an occasional error where an exception is thrown server-side in a race with a hole fill.  @Maithem, could you take a peek?  This appears to destabilize any concurrent updates to `SMRMap` objects.

The problem is quite non-deterministic, it's much easier to see when using the default logging level.  For detailed trace, reproduce using.

Server:

    % ./bin/corfu_server -Q -l /tmp/some/path -s 8000 --cm-poll-interval=99999 -d TRACE > & foo

Client, which performs random valid API ops from the `SMRMap` API.  All ops are using the same streamID and thus same `SMRMap` object.

    % cd src/test/erlang
    % ./Build.sh proper
    % ./Build.sh proper-shell
    > proper:quickcheck(proper:numtests(2000, map_qc:prop_parallel()), [noshrink]).

... and then wait for a `WARN` level message in the `foo` log file mentioning:

     17:50:21.760 ERROR [event-5] o.c.i.l.StreamLogFiles - Disk_write[0]: Exception
     org.corfudb.runtime.exceptions.OverwriteException: null
            at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)

Note that this usually happens at log offset 0, but I have seen it happen rarely at other offsets (e.g. 3 or 12).

At standard trace level, the output can look like this:

     17:50:21.524 WARN  [DistErl-8] o.c.i.SequencerServer - Sequencer recovery requested but checkpoint not set, defaulting to 0
     17:50:21.752 INFO  [event-5] o.c.i.l.StreamLogFiles - Opened new log file at /tmp/some/path/log0
     17:50:21.752 INFO  [event-11] o.c.i.l.StreamLogFiles - Disk_write[0]: Written to disk.
     17:50:21.760 ERROR [event-5] o.c.i.l.StreamLogFiles - Disk_write[0]: Exception
     org.corfudb.runtime.exceptions.OverwriteException: null
            at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    17:50:21.761 ERROR [event-5] o.c.i.NettyServerRouter - Exception during read!
     java.lang.RuntimeException: org.corfudb.runtime.exceptions.OverwriteException
            at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:213)
    Caused by: org.corfudb.runtime.exceptions.OverwriteException: null
            at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    17:50:21.768 INFO  [event-1] o.c.i.l.StreamLogFiles - Disk_write[1]: Written to disk.
     17:50:21.783 INFO  [event-15] o.c.i.l.StreamLogFiles - Disk_write[2]: Written to disk.
     17:50:26.758 WARN  [DistErl-1] o.c.r.v.AbstractView - Timeout executing remote call, invalidating view and retrying in 5s
     17:50:31.764 ERROR [event-5] o.c.i.l.StreamLogFiles - Disk_write[0]: Exception
     org.corfudb.runtime.exceptions.OverwriteException: null
            at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    17:50:31.764 ERROR [event-5] o.c.i.NettyServerRouter - Exception during read!
     java.lang.RuntimeException: org.corfudb.runtime.exceptions.OverwriteException
            at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:213)
    Caused by: org.corfudb.runtime.exceptions.OverwriteException: null
            at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    17:50:33.753 INFO  [event-13] o.c.i.l.StreamLogFiles - Disk_write[3]: Written to disk.

For a `TRACE` level log, see https://gist.github.com/slfritchie/eac869fd354eeda4a1bbc56db995cd57

I forgot to mention the client's behavior in all of this.  When the bug on the server side is triggered:

* Both threads are unable to perform more put() & get() ops

* The client logs these at 10 second intervals:

    01:11:00.417 WARN  [Thread-4] o.c.r.v.AbstractView - Timeout executing remote call, invalidating view and retrying in 5s
     01:11:10.426 WARN  [Thread-4] o.c.r.v.AbstractView - Timeout executing remote call, invalidating view and retrying in 5s
     01:11:20.432 WARN  [Thread-4] o.c.r.v.AbstractView - Timeout executing remote call, invalidating view and retrying in 5s
     01:11:30.437 WARN  [Thread-4] o.c.r.v.AbstractView - Timeout executing remote call, invalidating view and retrying in 5s
     01:11:40.443 WARN  [Thread-4] o.c.r.v.AbstractView - Timeout executing remote call, invalidating view and retrying in 5s

* The server logs these at 10 second intervals but offset from the client's complaints.

    01:10:55.417 ERROR [event-4] o.c.i.l.StreamLogFiles - Disk_write[646]: Exception
     org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:10:55.417 ERROR [event-4] o.c.i.NettyServerRouter - Exception during read!
     java.lang.RuntimeException: org.corfudb.runtime.exceptions.OverwriteException
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:213)
    Caused by: org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:11:05.422 ERROR [event-4] o.c.i.l.StreamLogFiles - Disk_write[646]: Exception
     org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:11:05.423 ERROR [event-4] o.c.i.NettyServerRouter - Exception during read!
     java.lang.RuntimeException: org.corfudb.runtime.exceptions.OverwriteException
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:213)
    Caused by: org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:11:15.429 ERROR [event-4] o.c.i.l.StreamLogFiles - Disk_write[646]: Exception
     org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:11:15.430 ERROR [event-4] o.c.i.NettyServerRouter - Exception during read!
     java.lang.RuntimeException: org.corfudb.runtime.exceptions.OverwriteException
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:213)
    Caused by: org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:11:25.435 ERROR [event-4] o.c.i.l.StreamLogFiles - Disk_write[646]: Exception
     org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:11:25.436 ERROR [event-4] o.c.i.NettyServerRouter - Exception during read!
     java.lang.RuntimeException: org.corfudb.runtime.exceptions.OverwriteException
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:213)
    Caused by: org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:11:35.439 ERROR [event-4] o.c.i.l.StreamLogFiles - Disk_write[646]: Exception
     org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)
    01:11:35.440 ERROR [event-4] o.c.i.NettyServerRouter - Exception during read!
     java.lang.RuntimeException: org.corfudb.runtime.exceptions.OverwriteException
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:213)
    Caused by: org.corfudb.runtime.exceptions.OverwriteException: null
    	at org.corfudb.infrastructure.log.StreamLogFiles.append(StreamLogFiles.java:208)

--- a/infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java
+++ b/infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java
@@ -348,7 +348,7 @@ public class StreamLogFiles implements StreamLog {
                 throw new OverwriteException();
             }
             log.info("Disk_write[{}]: Written to disk.", logAddress);
-        } catch (Exception e) {
+        } catch (IOException e) {
             log.error("Disk_write[{}]: Exception", logAddress, e);
             throw new RuntimeException(e);
         }

--- a/test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java
+++ b/test/src/test/java/org/corfudb/infrastructure/log/StreamLogFilesTest.java
@@ -55,8 +55,7 @@ public class StreamLogFilesTest extends AbstractCorfuTest {
         log.append(address0, new LogData(DataType.DATA, b));
 
         assertThatThrownBy(() -> log.append(address0, new LogData(DataType.DATA, b)))
-                .isInstanceOf(RuntimeException.class)
-                .hasCauseInstanceOf(OverwriteException.class);
+                .isInstanceOf(OverwriteException.class);
     }
 
     @Test

--- a/test/src/test/java/org/corfudb/runtime/clients/LogUnitClientTest.java
+++ b/test/src/test/java/org/corfudb/runtime/clients/LogUnitClientTest.java
@@ -20,6 +20,7 @@ import java.io.RandomAccessFile;
 import java.util.Collections;
 import java.util.Set;
 import java.util.UUID;
+import java.util.concurrent.ExecutionException;
 
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
@@ -80,7 +81,7 @@ public class LogUnitClientTest extends AbstractClientTest {
         client.write(0, Collections.<UUID>emptySet(), 0, testString, Collections.emptyMap()).get();
         assertThatThrownBy(() -> client.write(0, Collections.<UUID>emptySet(), 0,
                 testString, Collections.emptyMap()).get())
-                .isInstanceOf(RuntimeException.class)
+                .isInstanceOf(ExecutionException.class)
                 .hasCauseInstanceOf(OverwriteException.class);
     }
 
@@ -94,7 +95,7 @@ public class LogUnitClientTest extends AbstractClientTest {
                 .isEqualTo(DataType.HOLE);
 
         assertThatThrownBy(() -> client.write(0, Collections.<UUID>emptySet(), 0, testString, Collections.emptyMap()).get())
-                .isInstanceOf(RuntimeException.class)
+                .isInstanceOf(ExecutionException.class)
                 .hasCauseInstanceOf(OverwriteException.class);
     }
 
@@ -109,7 +110,7 @@ public class LogUnitClientTest extends AbstractClientTest {
                 .isEqualTo(DataType.DATA);
 
         assertThatThrownBy(() -> client.fillHole(0).get())
-                .isInstanceOf(RuntimeException.class)
+                .isInstanceOf(ExecutionException.class)
                 .hasCauseInstanceOf(OverwriteException.class);
     }
 

